### ISSUE TITLE
gRPC error

### ISSUE THREAD

**Issue by yarox24** [Velocidex/velociraptor#2372]
![image](https://user-images.githubusercontent.com/18016218/211260744-0f2dcdd6-13a5-45a3-9259-d86e6a15855e.png)

Due to this error when trying to browse "Collected" section for specific host, it's impossible.
Velociraptor version: 0.6.7-4

Exact error message:
```
Error: grpc: received message larger than max (6308810 vs. 4194304)
```

**Comment by scudette** [comment]
This happens because the artifact collects too much data within the same raw - The GUI is limited to 5mb in one transfer and it is trying to view the table 10 rows at a time. There is more than 5mb of json in the first 10 rows ! 

You should rework the artifact to make it more efficient. If you still need to data just export into a zip file and view the raw json manually. This is just a GUI limitation - likely the browser will crash as well if there was no grpc limit.

**Comment by scudette** [comment]
Can you share the artifact that was causing issues? We can probably refactor it to send less data. With the new preview feature it is not necessary to send large JSON any more.

**Comment by scudette** [comment]
In 0.6.8 you can use the `preview_upload` column type to preview arbitrary large data blobs which can be uploaded from the client in the usual way. So there is no longer any need to add too much data into the JSON rows.

**Comment by yarox24** [comment]
I had new information about this issue. 

We are using following structure for VQL plugins:

Profile which contains multiple blocks, and blocks could refer to anything, but some of the blocks are referring to Artifact.Windows.KapeFiles.Targets


Fragment of profile:
```
name: Custom.Windows.XXX.Profile.Extended_Supplementary

type: CLIENT

parameters:

- name: AllUsersOfficeCache
  description: For all users collect Excel Autosave Location, Powerpoint Autosave Location, Publisher Autosave Location, Word Autosave Location, Office Document Cache
  type: bool
  default: Y
  

...
  
sources:

- name: CollectionStartTime
  query: SELECT *, "CollectionStartTime" AS __Source FROM Artifact.Custom.Windows.XXX.Helper.CurrentTime()

# Identification
- name: Identification
  precondition: SELECT OS From info() where OS = 'windows'
  query: SELECT * FROM Artifact.Custom.Windows.XXX.Identification()


- name: AllUsersOfficeCache
  query: SELECT * FROM if(condition=AllUsersOfficeCache,
        then={ SELECT * FROM Artifact.Custom.Windows.XXX.Block.AllUsersOfficeCache(IncludeVSS=IncludeVSS)
        } )
        
...
```

Example of block:
```
name: Custom.Windows.XXX.Block.AllUsersOfficeCache
description: |
   KAPE AllUsersOfficeCache 

type: CLIENT

parameters:
- name: IncludeVSS
  description: Gather even more data by including Volume Snapshot Service
  type: bool
  default: N

precondition:
    SELECT OS From info() where OS = 'windows'

sources:
   - query: |
        LET VSSAge <= if(condition=IncludeVSS, then=50, else=0) 
        SELECT * FROM Artifact.Windows.KapeFiles.Targets(VSSAnalysisAge=VSSAge,UseAutoAccessor=true,        OfficeAutosave=true,OfficeDocumentCache=true)
```

It seems that Windows.KapeFiles.Targets is duplicated multiple times in this API:
`api/v1/GetClientFlows?client_id=C.d6411095bf73df96&start_row=0&rows=10&sort_direction=false`

which in result generate following error:
`{"code":8, "message":"grpc: received message larger than max (5406552 vs. 4194304)"}
`

What can we do to fix duplicates of multiple Windows.KapeFiles.Targets in JSON body?

**Comment by scudette** [comment]
Do you still see this issue with a current version? (e.g. 0.73 or 0.72.4?) 

It looks like the API is including the compiled request in the response which for the case of KapeFiles (or artifacts derived from it) it is very large.

I just checked with recent builds though and the request is not included.

**Comment by yarox24** [comment]
Version: 0.72.4

![image](https://github.com/user-attachments/assets/9e677548-e040-4937-afb0-11fe5b91af02)

![image](https://github.com/user-attachments/assets/f69ef263-78a6-432c-9400-cad1d2de10f1)

**Comment by scudette** [comment]
Thanks for confirming the version - it should not be doing this. Can you please attach the full json output?

You can obtain this with the dev console - open the network tab then right click and copy object, paste it in notepad for the full JSON response.


![image](https://github.com/user-attachments/assets/5f08b8a2-ebda-4fe6-8a4f-0a04eb70eff1)

**Comment by yarox24** [comment]
For following query:
`api/v1/GetClientFlows?client_id=C.d6411095bf73df96&start_row=0&rows=4&sort_direction=false
`
{"code":8, "message":"grpc: received message larger than max (6049498 vs. 4194304)"}

So i'm sending anonymized output from: 
`api/v1/GetClientFlows?client_id=C.d6411095bf73df96&start_row=0&rows=3&sort_direction=false
`

[GetClientFlowsAnon.zip](https://github.com/user-attachments/files/17378301/GetClientFlowsAnon.zip)
I would like to remove this file after few days.

**Comment by scudette** [comment]
Thank you so much for including this data - It helps a lot!

I did not realize that the collection came from a hunt - I was trying to replicate with directly issuing a collection. When a hunt is created, it pre-compiles the artifact into the collection request in a field called "compiled_collector_args". This is done so it does not need to recompile the same thing for each client in the hunt (there could be thousands of clients in a hunt). 

The compiled_collector_args contains the entire request that will be sent to the client so in the case of the KapeFiles.Target artifact it is very large. While this data is useful to the hunt manager it is not really needed for the GUI and it is discarded in fact. But this can add up a lot as you found and it exceeds the API size limits.

I will prepare a fix shortly

**Comment by scudette** [comment]
Fixed in 0.73.2
